### ISSUE TITLE
New type errors in Model.create with version 7.0.0

### ISSUE THREAD

**Issue by coolsoftwaretyler** [mobxjs/mobx-state-tree#2233]
**_Bug report_**

-   [x] I've checked documentation and searched for existing issues [and discussions](https://github.com/mobxjs/mobx-state-tree/discussions)
-   [x] I've made sure my project is based on the latest MST version
-   [x] Fork [this](https://codesandbox.io/s/y64pzxj01) code sandbox or another minimal reproduction.

Originally reported in the v7.0.0 discussion post: https://github.com/mobxjs/mobx-state-tree/discussions/2224#discussioncomment-11550012

**Sandbox link or minimal reproduction code**

Here is a CodeSandbox reproduction.

On v6.0.1, we get a type error correctly: https://codesandbox.io/p/sandbox/naughty-bassi-forked-39c7cz?file=%2Fsrc%2Findex.ts&workspaceId=ws_QGa4cGacuoY7BgaNFKGDj8

![Screenshot 2024-12-13 at 12 31 05 PM](https://github.com/user-attachments/assets/ef59c1a4-6e36-43de-a1d7-dd7b8ca33254)

On v7.0.0, MST will fail at runtime, but TypeScript does not warn us about that problem: https://codesandbox.io/p/sandbox/gv3tn2?file=%2Fsrc%2Findex.ts

![Screenshot 2024-12-13 at 12 31 45 PM](https://github.com/user-attachments/assets/1c672bfe-1a2c-46cc-b6c2-c909558d6692)

**Describe the expected behavior**

We should receive type errors on the `a` property when passing a string to a `types.number`.

**Describe the observed behavior**

No type error in version 7.0.0

**Comment by thegedge** [comment]
Let me take a shot at fixing the mess I keep creating 😂 🙈

**Comment by thegedge** [comment]
I created a regression test here
![CleanShot 2024-12-14 at 15 08 29@2x](https://github.com/user-attachments/assets/1ab8bdcb-1286-4bff-b595-3ff9a833099b)

Looks like it failed. I traced it to the version of TypeScript. Seems like the typechecking error shows up again in >=4.6.2. Is there a minimum TS we want to support, or should we try to make it work for <4.6.2?

**Comment by coolsoftwaretyler** [comment]
> Let me take a shot at fixing the mess I keep creating

Haha, it's *our* mess, my friend.

The typescript version makes a ton of sense here, because I don't think I've seen this on apps I'm working on with MST 7 and TypeScript 5.

I haven't thought enough about TS version requirements yet. If there's an easy path for supporting an older version, it's worth it. But if you find it's taking too long to fix or is gonna spur more headaches, I think we can evaluate what a good TS target is. We haven't put thought into that for a while here, so it's worth consideration.

**Comment by thegedge** [comment]
👍 sounds good. I'll pick at it, see if I can't make something work in both versions.

I'll see if I can set up some kind of TS matrix to test older versions. I don't think it's worth going lower than 4, and I'll see the lowest minor version of 4 I can easily get working.

Once we figure out the min version, what do you think about using peer dependencies (and [`peerDependenciesMeta`](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#peerdependenciesmeta) to make TS optional):

```json
{
  "peerDependencies": {
    "typescript ">=4.x.y",
  },
  "peerDependenciesMeta": {
    "typescript": {
      "optional": true
    }
  }
}
```

That way people will know any older version of TS is probably not gonna work. Looks like it was introduced in a [later minor version of npm 6](https://github.com/npm/cli/issues/1247). That's more than 4 years old, so I'd say reasonable to expect at least that, eh?

**Comment by coolsoftwaretyler** [comment]
Yeah I feel good about that as a fallback!

**Comment by tm-3** [comment]
It's a little more than the typescript version. I'm using 5.7.2 and getting typing errors. I thought I was crazy for a bit because I could have sworn that Model.create() used to infer the type properly

 Here's what I am seeing:

- TS =5.7.2
- MST > 7.0.0

Model.create does not properly type the create() method of the ModelType. This goes for maps, arrays, and models. 

- TS = 5.7.2
- MST 6.0.1 

As stated above everything works normally with MST 6. 

The big difference is the use of ts-essentials in v7. I removed teh 

From this:

    /**
     * Creates an instance for the type given an snapshot input.
     *
     * @returns An instance of that type.
     */
    create(snapshot?: C | ExcludeReadonly<T>, env?: any): this["Type"];

to this:

```
    /**
     * Creates an instance for the type given an snapshot input.
     *
     * @returns An instance of that type.
     */
    create(snapshot?: C , env?: any): this["Type"];
```
Everything worked as expected here with latest versions. Then I installed ts-essentials as a dependency while I was messing around with the ExcludeReadonly<T> and the use of WritabelKeys<T> from ts-essentials and noticed everything worked as expected. It's probably just not including the ts-essentials library when building declarations or something simple for someone who has more patience than I do with TS. 

hth.

**Comment by coolsoftwaretyler** [comment]
@tm-3, that's super helpful, thank you!

I'm about to be on a cross country move, but by early 2025 I can look deeper into this and see what else we can do to fix things.

Sorry for the inconvenience!

**Comment by tm-3** [comment]
not a problem at all. Adding ts-essentials as a dependency works for now. If I get extra time I'll dig a bit deeper, but probably not until after Christmas. 

Safe travels!

**Comment by thegedge** [comment]
Appreciate the digging deeper, @tm-3. I've been on vacation and now super busy prepping for the holidays, but I should have some time to look at this before the new year.

It's too bad that all of these changes have caused regressions, but we can capture all of these in typechecking tests and prevent it from happening again 🙌

**Comment by OnkelTem** [comment]
@tm-3 I was also thinking I'm going crazy when I couldn't replicate locally a `model.create()` error that should have been there according to a [sandbox here](https://www.typescriptlang.org/play/?target=6&jsx=0#code/JYWwDg9gTgLgBAbzvAvnAZlCI4CIQQBGAHgLQDOMAhjAKakxS224DcAUOwMYQB2lyAO7N4AXmQA6AgBNaAGwAUCdnDiEI0gJ4AuSZSjBeAcwA07FAEoO7GMNowJXJjVpK4VXQEY4ljkA

):

```ts
import { t } from "mobx-state-tree";

const tweet = t.model({
  body: t.string,
});

tweet.create({ a: 1 });
```

![Image](https://github.com/user-attachments/assets/e4592384-7c48-4fea-8527-442b07334886)

Thanks @mkantor for investigating this issue and finding out it's the `ts-essentials` package that was missing.

And yeah, this should be addressed, because imagine - new devs would think that `model.create()` is unable to check parameters, as it happened with me one hour ago, until I accidentally figured out it actually works - thanks to the TypeScript Playground which presumably installs deps in its own way which... differs from how npm gets it done.

**Comment by thegedge** [comment]
Thanks for looking into this and figuring out the problem, @mkantor. My apologies for not getting to it sooner (and making this breaking change 😢).

I'll put together a PR right away and we'll get a release out soon!

**Comment by coolsoftwaretyler** [comment]
Thanks to everyone helping out here! I just merged #2245 and opened a follow-up issue so we can eventually remove the dependency (without TS bugs) over in https://github.com/mobxjs/mobx-state-tree/issues/2246.

I'll push this as a patch version. Sorry for adding a dep to everyone's bundles, it should hopefully be temporary. :)

**Comment by coolsoftwaretyler** [comment]
Fixedin version `7.0.2` which is available on npm now.
